src/Components/Server/src/ComponentHub.cs

```diff
@@ -20,17 +22,23 @@ internal sealed class ComponentHub : Hub
         private static readonly object CircuitKey = new object();
         private readonly CircuitFactory _circuitFactory;
         private readonly CircuitRegistry _circuitRegistry;
+        private readonly CircuitOptions _options;
         private readonly ILogger _logger;
 
         /// <summary>
         /// Intended for framework use only. Applications should not instantiate
         /// this class directly.
         /// </summary>
-        public ComponentHub(IServiceProvider services, ILogger<ComponentHub> logger)
+        public ComponentHub(
+            CircuitFactory circuitFactory,
+            CircuitRegistry circuitRegistry,
+            ILogger<ComponentHub> logger,
+            IOptions<CircuitOptions> options)
```

Comment:
In an earlier PR, @rynowak was making the point that adding more and more constructor params here is going to spiral out of control in the future, as we take dependencies on more and more DI services. The alternative would be continuing to take `IServiceProvider` and doing `GetRequiredService`.

Am I representing your point correctly, @rynowak?

Reply:
A few points:
* I'm not 100% sure the number of parameters will increase over what there is currently here.
  * There are two types that are essential to the functionality of this class:
     * CircuitFactory to create CircuitHosts
     * CircuitRegistry to handle state transitions.
  * Options and logging are mostly for cross-cutting concerns.
 * I would expect every method other than Connect/StartCircuit to be a one-liner that dispatches to the CircuitHost.
   * Its an internal type, so adding more stuff to the constructor doesn't have any negative impact.
   * By having the dependencies injected on the constructor we have more visibility into what this class needs to work and its easier to realize when its doing too much.

I particularly dislike injecting the IServiceProvider unless there is a good reason for it as it hides dependencies from plain inspection, while this way, just looking at the constructor I can know what is going on.

Reply:
Ah yes it is internal so I don't really feel strongly. I may even have been misremembering which type @rynowak was commenting about before, but he can say if so.

Reply:
Yeah I was talking about public types. This seems right.